src/handlers.cpp

```diff
@@ -1579,28 +1632,129 @@ void ImpuRegDataTask::on_sar_response(Diameter::Message& rsp)
         TRC_DEBUG("Getting User-Data from SAA for cache");
         saa.user_data(_xml);
         put_in_cache();
+        caching = true;
       }
-      send_reply();
       break;
     case 5001:
     {
       TRC_INFO("Server-Assignment answer with result code %d - reject", result_code);
       SAS::Event event(this->trail(), SASEvent::REG_DATA_HSS_FAIL, 0);
       SAS::report_event(event);
-      send_http_reply(HTTP_NOT_FOUND);
+      _http_rc = HTTP_NOT_FOUND;
     }
     break;
     default:
       TRC_INFO("Server-Assignment answer with result code %d - reject", result_code);
       SAS::Event event(this->trail(), SASEvent::REG_DATA_HSS_FAIL, 0);
       SAS::report_event(event);
-      send_http_reply(HTTP_SERVER_ERROR);
+      _http_rc = HTTP_SERVER_ERROR;
       break;
   }
-  delete this;
+
+  // If we're not caching the registration data, check for deregistration and
+  // delete the task now
+  if (!caching)
+  {
+    // Even if the HSS rejects our deregistration request, we should
+    // still delete our cached data - this reflects the fact that Sprout
+    // has no bindings for it.
+    bool deleting = false;
+    if (is_deregistration_request(_type))
+    {
+      SAS::Event event(this->trail(), SASEvent::REG_DATA_HSS_SUCCESS, 0);
+      SAS::report_event(event);
+      std::vector<std::string> public_ids = XmlUtils::get_public_ids(_xml);
+      if (!public_ids.empty())
+      {
+        TRC_DEBUG("Got public IDs to delete from cache - doing it");
+        for (std::vector<std::string>::iterator i = public_ids.begin();
+             i != public_ids.end();
+             i++)
+        {
+          TRC_DEBUG("Public ID %s", i->c_str());
+        }
+
+        SAS::Event event(this->trail(), SASEvent::CACHE_DELETE_IMPUS, 0);
+        std::string public_ids_str = boost::algorithm::join(public_ids, ", ");
+        event.add_var_param(public_ids_str);
+        std::vector<std::string> associated_private_ids = get_associated_private_ids();
+        std::string associated_private_ids_str = boost::algorithm::join(associated_private_ids, ", ");
+        event.add_var_param(associated_private_ids_str);
+        SAS::report_event(event);
+        CassandraStore::Operation* delete_public_id =
+          _cache->create_DeletePublicIDs(public_ids,
+                                         associated_private_ids,
+                                         Cache::generate_timestamp());
+        CassandraStore::Transaction* tsx = new CacheTransaction(this,
+                                        &ImpuRegDataTask::on_del_impu_success,
+                                        &ImpuRegDataTask::on_del_impu_failure);
+        _cache->do_async(delete_public_id, tsx);
+        deleting = true;
+      }
+    }
+
+    // If we're not deleting the public IDs, reply and quit now.
+    if (!deleting)
+    {
+      if (_http_rc == HTTP_OK)
+      {
+        send_reply();
+      }
+      else
+      {
+        send_http_reply(_http_rc);
```

Comment:
Should we delegate this _http_rc test to send_reply to avoid duplicating it everywhere?